### DISCUSSION TITLE
TypeError: Cannot read property 'tagName' of null

### DISCUSSION THREAD

**Discussion by piyu-sh** [vercel/next.js#12830] [Help]
So I get this error
![image](https://user-images.githubusercontent.com/7107647/81789123-27a7f700-9521-11ea-95f5-efd47cd51e9a.png)

I found this relevant comment.
https://github.com/zeit/next.js/issues/8220#issuecomment-575487571

Because I'm using next/head in _app.tsx like this 

```
export default class MyApp extends App {
  render(): JSX.Element {
    const { Component, pageProps } = this.props;
    return (
      <>
        <Head>
          <NewRelicSnippet />
        </Head>
        <Component {...pageProps} />
      </>
    );
  }
}
```

When I remove this `Head` tag, this error is gone, but I need it to inject new relic script. Can anyone help please ?

**Comment by jamesmosier** [comment]
Could you share what is inside `NewRelicSnippet`?

I was able to run this with no issues:

`pages/_app.js`
```jsx
import Head from 'next/head';
export default function MyApp({ Component, pageProps }) {
  return (
    <>
      <Head>
        <title>My page title</title>
      </Head>
      <Component {...pageProps} />
    </>
  );
}
```

as well as this inside the `<Head>` tags in `_app`:

```jsx
<script
  dangerouslySetInnerHTML={{
    __html: `
      (function(w,d,s,l,i){w[l]=w[l]||[];w[l].push({'gtm.start':
      new Date().getTime(),event:'gtm.js'});var f=d.getElementsByTagName(s)[0],
      j=d.createElement(s),dl=l!='dataLayer'?'&l='+l:'';j.async=true;j.src=
      'https://www.googletagmanager.com/gtm.js?id='+i+dl;f.parentNode.insertBefore(j,f);
      })(window,document,'script','dataLayer', 'asdf');
    `,
  }}
/>
```

**Reply by piyu-sh** [reply]
This is how `NewRelicSnippet` looks 
```
const NewRelicSnippet = (): JSX.Element => {
  return (
    <script
      type="text/javascript"
      dangerouslySetInnerHTML={{ __html: snippet }}
    />
  );
};
```
And `snippet` is script for New relic browser agent.
And one thing this is behaviour in Next@9.4.

**Reply by jamesmosier** [reply]
Do you have other code you could share that shows this issue happening? I wasn't able to reproduce so far with what you've shared.

**Reply by puneetv05** [reply]
@jamesmosier  I also have this issue but in my case, I need to inject payment gateway script :(

**Reply by jamesmosier** [reply]
You need a few things in order to have others help solve this 😄 

* Open up a new Discussion with full instructions on how to reproduce this.
* Provide a full repository with the exact code that is breaking. Otherwise others have to create a new project, try and guess all the things about your code and setup, and then hopefully reproduce it.

**Comment by tejas-hosamani** [comment]
~~Same here. I am facing the same exact issue. I've a script tag inside Head tag like so:~~

```jsx
<Head>
      <title>Page title</title>
      <script src="https://js.instamojo.com/v1/checkout.js" defer></script>
</Head>
```

Edit: False alarm.

**Reply by puneetv05** [reply]
hey @CodingCreate101  did you created an Issue regarding this? or should we create one? Facing the same issue but with different script.

**Reply by tejas-hosamani** [reply]
False alarm. I just figured it wasn't because of script tag. It's because of how I am handling callback function. Still investigating. @puneetverma05

**Comment by puneetv05** [comment]
> @jamesmosier I also have this issue but in my case, I need to inject payment gateway script :(

👎 Its been 21 days no one here to respond :(

**Reply by jamesmosier** [reply]
You need a few things in order to have others help solve this 😄 

* Open up a new Discussion with full instructions on how to reproduce this.
* Provide a full repository with the exact code that is breaking. Otherwise others have to create a new project, try and guess all the things about your code and setup, and then hopefully reproduce it.

**Reply by tejas-hosamani** [reply]
@puneetverma05 Try to reverse engineer, see when the error occurs.

**Reply by puneetv05** [reply]
@CodingCreate101  thanks will do, what was the issue? how did you fix it?  my case, the error occurs in head manager. will definitely start a new discussion as @jamesmosier mentioned with detailed info about the error.

**Reply by tgrecojs** [reply]
@puneetverma05 did you start a new thread? 

I just had this issue arise in 2 seperate next.js projects. Whenever I try to integrate next.js with tailwind-css I always run into an issue. I've been lucky enough to get past it on a few different occasions but I'm not exactly sure how as resolving this always forces me to re-create a project from scratch. Once I have everything going then I'll make note of the repo in the event that I need to scaffold out a project with tailwind in the future, thus why I created this [nextjs-web3-starter repo](https://github.com/tgrecojs/nextjs-web3-starter).

It's been about 6 months since I used the project as a boilerplate though and in that time it seems some changes have led to this error arising for me yet again. Again, **this issue was introduced into my project after adding talwind-css**. It makes sense that it's related to styled components built with tailwind. **More specificially, this error is a side-effect of trying to leverage the tailwind.macro library via babel's macros plugin.** I'm not sure where in the pipeline the error is occuring, but [this branch of my nextjs-web3-starter](https://github.com/tgrecojs/nextjs-web3-starter/tree/video-3-transaction) is the latest next.js + tailwind.macro codebase to become victim to breaking changes after a few months of stagnation. 

I'm confident I'll get through it so i'm not going to open up a new thread (unless someone thinks it would be wholly beneficial for others in the community).

**Reply by puneetv05** [reply]
Hey @tgrecojs , I absolutely agree that there is no point in starting a new separate thread for a similar issue without providing insights into what actually the problem is. I'll reproduce the error on a sample repo hopefully by the weekend I'll start a thread with relevant information so that it would be meaningful for the contributors and community,
 btw I'm also using the tailwind-css in the project but my issue is with injecting a payment gateway script in head, getting `headcount error with tagname of null`

**Reply by tgrecojs** [reply]
@puneetverma05 I managed to get my issue sorted 🥳. I'm sorry to say that I have no new knowledge as to why this was occurring though as I resolved it by fixing conflicting contents inside my `package-lock.json` file. I can say for a fact that adding styling packages like tailwind.maco, emotion-js, etc. is what caused me to run into this issue today in my project. This is why I tried to clone my old repository - I was just going to migrate my code into it, but then I saw the same `tagName` issue.

As i mentioned in my last comment, I knew that I had it working properly on my one branch. Before I ran my code from that branch, I resolved a merge conflict without looking into the package-lock.json which is where the conflict was occurring. Once i saw that both my master branch as well as the video-3 branch were both throwing the error, I figured it would be worth trying to merge in the video-3 branch to master. Only this time, I took a gamble and resolved the merge conflict by only using the contents from the video-3 branches `package-lock.json` file. 

I've linked to the file below in case that you might be running into issues with the same dependencies that I was. As suspected, they were all babel related.

https://github.com/tgrecojs/nextjs-web3-starter/commit/2052ee739861f2e921e590b8176980010352ff26#diff-32607347f8126e6534ebc7ebaec4853d

Best of luck in getting this resolved!

### TL;DR
Fixed this issue by updating my package-lock.json file to target updated versions of (for the most part) strictly babel-related dependencies.

**Reply by tejas-hosamani** [reply]
> Hey @tgrecojs , I absolutely agree that there is no point in starting a new separate thread for a similar issue without providing insights into what actually the problem is. I'll reproduce the error on a sample repo hopefully by the weekend I'll start a thread with relevant information so that it would be meaningful for the contributors and community,
> btw I'm also using the tailwind-css in the project but my issue is with injecting a payment gateway script in head, getting `headcount error with tagname of null`

@puneetverma05 
I encourage you to confirm whether that's really the case -> Whether the issue is due to adding a script tag. My issue was/is relevant to how I was handling callbacks. I thought it was because of the script tag too, but it wasn't. Just remove/undo some payment-related code, remove some callbacks and try WITH script tag and see if it still occurs.

**Reply by saranshgupta1995** [reply]
Had been trying to integrate with Razorpay when I ran into this.
While debugging, I saw that it only happens if I have a handler function ( a callback ) to add while triggering Razorpay.
Will be adding a sample code in a bit.

**Reply by puneetv05** [reply]
@saranshgupta1995 Yes, My Payment Gateway Script is of Razorpay https://github.com/vercel/next.js/discussions/12830#discussioncomment-38115

**Reply by puneetv05** [reply]
> > Hey @tgrecojs , I absolutely agree that there is no point in starting a new separate thread for a similar issue without providing insights into what actually the problem is. I'll reproduce the error on a sample repo hopefully by the weekend I'll start a thread with relevant information so that it would be meaningful for the contributors and community,
> > btw I'm also using the tailwind-css in the project but my issue is with injecting a payment gateway script in head, getting `headcount error with tagname of null`
> 
> @puneetverma05
> I encourage you to confirm whether that's really the case -> Whether the issue is due to adding a script tag. My issue was/is relevant to how I was handling callbacks. I thought it was because of the script tag too, but it wasn't. Just remove/undo some payment-related code, remove some callbacks and try WITH script tag and see if it still occurs.

RIght, issue is not with the script tag its with the callback, If I remove the callback the issue is not present anymore.

**Reply by harikrishnan-neibre** [reply]
@puneetverma05 could you sort this issue with Razorpay PG. Could you share how you managed if you have sorted it

**Comment by lolobosse** [comment]
For having the same issue; I fixed it by removing the empty list that I had:

I injected: 

```
 <>
        {page > 1 && <link rel="prev" href={`https://${domain}${createLink({what, where, page: page - 1})}`}/>}
        {page < total_pages &&
        <link rel="next" href={`https://${domain}${createLink({what, where, page: page + 1})}`}/>}
    </>
```
Which is an empty if I have only one page so it was creating an error; fix, in my case:

```page > 1 ? <>
        {page > 1 && <link rel="prev" href={`https://${domain}${createLink({what, where, page: page - 1})}`}/>}
        {page < total_pages &&
        <link rel="next" href={`https://${domain}${createLink({what, where, page: page + 1})}`}/>}
    </> : null
```

I do not know if you can consider that there is a bug in the lib if a list (which was itself in a list) is empty. 

Consider this code:

```
export default function IndexPage() {
  return (
    <div>
      Hello World.{" "}
      <Head>
        <>
          <script src="test" />
          <></>
        </>
      </Head>
      <Link href="/about">
        <a>About</a>
      </Link>
    </div>
  );
}
```

This code is throwing the error that @piyu-sh was mentionning.

Have a look at the [sandbox](https://codesandbox.io/s/fervent-antonelli-vm3ee?file=/pages/index.js:61-321) I made.

**Comment by auplopwar** [comment]
Any update on this issue, I am seeing this when a list of elements is returned from a component and that component is added to `<Head />`

e.g.  
```
const ExtensionScripts = () => { 
    return (
       <>
           <script src='' />
           <script src='' />
       </>
    )
}

const App = () => {
    return (
       <Head>
           <ExtensionScripts />
       </Head>
    )
}
```

**Reply by lbarthon** [reply]
I'm having the same error 😢

**Comment by thiago-diniz** [comment]
I have same problem

**Comment by zulnam** [comment]
TLDR: To anyone having this issue, check your Head component and progressively remove scripts, links, styles; mostly everything. Eventually you'll find the line causing the error.

Recently came over this issue and began debugging to see what was causing it.

The problem was a customised style tag in the next Head component, using jsx global.
Bad Code:
```
<style jsx global>{`
        * {
          font-family: 'Ubuntu Mono', monospace;
          background-color: black;
          color: white;
        }
        body {
          padding: 0;
        }
      `}</style>
```

I moved the global styles outside of the Head  To make it look like this:
```
<>
      <Head>
        <title>Wouldn't you like to know, weather boy.</title>
        <meta name="viewport" content="initial-scale=1.0, width=device-width" />
        <link
          href="https://fonts.googleapis.com/css?family=Ubuntu+Mono&display=swap"
          rel="stylesheet"
        />
    </Head>
    <Global styles={globalStyles} />
</>
```

I use emotion to style my app (the poorly named library, not the psychological state), so I moved the styling in a separate file and imported Global from "@emotion/core".

Hope this helps.

**Comment by sentisso** [comment]
Same error occurs when I simply have the next/script `<Script>` element in `<Head>`. Just like this:
```
// _app.tsx
import Script from 'next/script'

<Head>
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <link rel="icon" href="/favicon.ico" />
    <Script strategy="afterInteractive" src="https://cdn.jsdelivr.net/npm/bootstrap@5.0.2/dist/js/bootstrap.min.js" />
</Head>
```
When I remove the Script tag or when I use the native `<script>` tag, the error dissapears.

**Comment by awitherow** [comment]
I am getting this since I started using `next-seo`.

my Layout file that I use the `<Head />` in looks like this...

```javascript
      <Head>
        <link rel="icon" href="/favicon.ico" />
      </Head>

      <NextSeo title={`${title} | ${subtitle}`} description={description} />
```

**Comment by baotlake** [comment]
I have same problem. finally, I found out it was because I sued A empty `React.Fragment`(`<></>`). And solved the problem after I replaced `<></>` with `<meta></meta>`.

**Comment by xbryan813x** [comment]
According to https://nextjs.org/docs/basic-features/script we cant add scripts to the `<Head>` component, the preferred way of adding scripts is in the body with the 'next/script' component. You can add a prop value indicating at what point of the page render cycle youd like to initialize it. 

For Example my `_app.js` component looks like 

function MyApp({ Component, pageProps }) {
  return (
    `<>`
      `<Head>`
        `<link rel="icon" href="/favicon.ico" />`
      `</Head>`
      `<TrackingScripts />`
      `<Nav />`
      `<Component {...pageProps} />`
      `<Footer />`
    `</>`
  );
}

the script im including looks like

function Tracking() {
  return (
    `<Script`
      `strategy="afterInteractive"`
     ` dangerouslySetInnerHTML={{`
       ` __html: `script source code`,`
     ` }} />`
  );
}

Hope that helps!

**Reply by sodhisaab** [reply]
Good, it works now. I missed that we can not use Script into Head component 🤦🏼

**Comment by HurmisIndia** [comment]
I was adding a <style jsx> in <Head> which gave me same error. 
while css-in-js <style> component should be in <body> tag.

**Comment by mydearxym** [comment]
in my case, i put Script in Head cause the error: 

wrong: 

```jsx
import Head from 'next/head'
import Script from 'next/script'

      <Head>
        <meta charSet="utf-8" />
        <meta name="renderer" content="webkit" />
   
       <Script
        src={`https://www.googletagmanager.com/gtag/js?id=${GA.TRACKING_ID}`}
      />

      </Head>

    ...
```

right: 

```jsx
import Head from 'next/head'
import Script from 'next/script'

      <Head>
        <meta charSet="utf-8" />
        <meta name="renderer" content="webkit" />
  
      </Head>

       <Script
        src={`https://www.googletagmanager.com/gtag/js?id=${GA.TRACKING_ID}`}
      />

    ...
```

**Comment by lucagonzalez** [comment]
```
<div className="page-wrapper">
            <Head>
                <title>Sign Up</title>
                <meta name="description" content="Signup page"/>
                <link rel="icon" href="/favicon.ico"/>
            </Head>
            <div className="signup-page pb-10">
                <div className="w-full max-w-siteContainer mx-auto px-4">
                    <div className="page-head flex justify-between items-center py-6">
                        <div className="title">
                            <h1 className="text-xl lg:text-2xl text-DarkGray font-bold Quicksand">
                                Sign Up
                            </h1>
                        </div>
                    </div>
                </div>
            </div>
</div>
```

For me this is causing the same problem. I tried removing everything. Just the presence of the ` <Head> ` tag breaks my code

**Comment by FlorianWendelborn** [comment]
I had to put the `<Script>` into a `<div>` when trying to use it inside `Layout.tsx`...
